tests/test_xbundle.py

```diff
@@ -0,0 +1,155 @@
+
+
+class TestXBundle(TestCase):
+    """
+    Misc tests for XBundle object.
+    """
+
+    def test_set_course(self):
+        """
+        Test functionality of set_course.
+        """
+        input_xml = input_data.EMPTY_COURSE
+
+        bundle = XBundle(keep_urls=True)
+        bundle.load(file_from_string(input_xml))
+
+        # No org or semester is specified in XML above.
+        self.assertEqual(bundle.course.get("org"), None)
+        self.assertEqual(bundle.course.get("semester"), None)
+        self.assertEqual(bundle.semester, "")
+
+        # Note lack of org attribute and url_name for course element.
+        course_str = input_data.NO_COURSE
+        with self.assertRaises(Exception) as ex:
+            bundle.set_course(etree.XML("<x>" + course_str + "</x>"))
+        self.assertTrue(
+            "set_course should be called with a <course> element"
+            in ex.exception.args)
+
+        with self.assertRaises(Exception) as ex:
+            bundle.set_course(etree.XML("<course />"))
+        self.assertTrue("No semester found." in ex.exception.args)
+
+        bundle.set_course(etree.XML("<course url_name='x' />"))
+        self.assertEqual(bundle.semester, "x")
+
+        bundle.set_course(etree.XML(course_str))
+
+        # MITx is not present in data, it is automatically set.
+        self.assertEqual(bundle.course.get("org"), "MITx")
+        self.assertEqual(bundle.course.get("semester"), "2013_Spring")
+        self.assertEqual(bundle.semester, "2013_Spring")
+
+        bundle_string = str(bundle)
+
+        expected = expected_data.SET_COURSE
+        self.assertEqual(clean_xml(bundle_string), clean_xml(expected))
+
+    def test_add_descriptors(self):
+        """
+        Test add_descriptors.
+        """
+        # Note url_name_orig in chapter.
+        input_xml = input_data.URL_NAME_ORIG_IN_CHAPTER1
+        bundle = XBundle(keep_urls=True)
+        bundle.load(file_from_string(input_xml))
+
+        # str(bundle) doesn't change input xml, but export_to_directory will.
+        self.assertEqual(clean_xml(input_xml), clean_xml(str(bundle)))
+
+        old_current_dir = os.getcwd()
+        tempdir = mkdtemp()
+        try:
+            os.chdir(tempdir)
+            bundle.export_to_directory()
+
+            bundle2 = XBundle(keep_urls=True)
+            bundle2.import_from_directory()
+
+            expected = expected_data.URL_NAME_ORIG
+
+            self.assertEqual(clean_xml(expected), clean_xml(str(bundle2)))
+        finally:
+            os.chdir(old_current_dir)
+            rmtree(tempdir)
+
+    def test_is_not_random_urlname(self):
+        """
+        Test behavior of is_not_random_urlname.
+        """
+        # Randomness test used in method
+        input_hash = 'z5bc076ad06e4ede9d0561948c03be2f'
+        input_letters = 'zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz'
+        input_empty = ''
+
+        # Since keep_studio_urls is True
```

Comment:
This could use an additional comment, because this test doesn't make any sense on face value.

The test is correct -- but you have to go read the function it's testing to understand why this isn't crazy.

Suggestion: "Function always returns `True` if `self.keep_studio_urls` is True.